Setup and define user stories in README

marlisebriedenhann opened the pull request:
### Hangman Brief
The purpose of this project is to:
Teach the fundamentals of Ruby
Teach the fundamentals of git
Teach the fundamentals of TDD
Introduce basic OO design concepts

The technical requirements:
Ruby
RSpec for unit testing
Git version control
No other gems
CLI only

Although we have no way of enforcing this, it’d be a good exercise to write tests first before implementing anything; at least on a per object basis.
Upfront design is also highly encouraged, thinking about which objects might exist in this program and their responsibilities.

The User stories
As a player, I want to start a game of Hangman
As a player, I want to see the characters I’ve guessed correctly in a word
As a player, I want to see the characters I have guessed incorrectly
As a player, I want to see how many lives I have left
As a player, I want to see a victory message when I have won
As a player, I want to see a message when I have lost, including the word that I didn’t guess
As a player, I want to be able to play again with a different word so that I can play many times


### Changes
- add feature to enable a player to start a game of hangman

zubin commented on README.md:
```diff
@@ -0,0 +1,13 @@
+When the game starts, it generates a random word.  The random word is masked before displaying on screen.  The game continues until the number of attempts is exhausted or the masked word is revealed by the user successfully guessing all masked characters.
+
+### The User stories
+As a player, I want to start a game of Hangman
```
Perhaps it's worth documenting how each (or some) of these look, eg this one could be:

```sh
$ script/hangman
```

marlisebriedenhann replied:
👍

zubin commented on spec/hangman_spec.rb:
```diff
@@ -0,0 +1,14 @@
+# frozen_string_literal: true
+
+require './lib/hangman'
```
Normally the `lib` directory is added to project's `$LOAD_PATHS` (this is where ruby looks for `require`d files). 

Therefore, this should work:

```suggestion
require 'hangman'
```

zubin commented on spec/hangman_spec.rb:
```diff
@@ -0,0 +1,14 @@
+require './lib/hangman'
+
+RSpec.describe Hangman do
+  describe ".start" do
```
Normally this is the method under test (sometimes the `subject`; or called on the `subject`), but it's not used in the test.

zubin replied:
If, as this `describe` text suggests, the goal of this PR is to fulfil the first user story:

> As a player, I want to start a game of Hangman

Then we should define a "done-when" for it. This BDD/TDD approach is great because it forces/encourages us to be outcome-oriented and focussed, rather than diving into code (somewhat randomly, as we developers are prone to do!). 

In this case, at least for me, it's hard to decide what this done-when looks like without taking a step back to see the bigger picture (where does `Hangman.start` get used?; what does the rest of the script look like?)

As suggested earlier, IMO scripts should be as simple as possible (ie leave the heavy lifting and logic to testable objects in the codebase). with this in mind the script _could_ simply invoke `Hangman.start`. But that leaves us with the complexity of testing keyboard interactions. Let's discuss some alternative approaches.

zubin commented on README.md:
```diff
@@ -0,0 +1,13 @@
+As a player, I want to see how many lives I have left
+As a player, I want to see a victory message when I have won
+As a player, I want to see a message when I have lost, including the word that I didn’t guess
+As a player, I want to be able to play again with a different word so that I can play many times
```
Consider using checkboxes to indicate which stories are complete. This can be achieved by inserting `* [ ] ` before each line (or `* [x] ` if it's complete).

zubin reviewed (commented):
Great start! 👏 

May I suggest restricting this PR to:

* RSpec boilerplate
* Defining user stories
* RDD (let's discuss)

(ie remove script and spec files from this PR)

Then we can focus on each user story afterwards. WDYT?

marlisebriedenhann committed:
Fix typo on README, follow RDD

marlisebriedenhann committed:
Initialise rspec

zubin commented on README.md:
```diff
@@ -0,0 +1,65 @@
+# README
```
The main heading is usually the name of the application or library.

marlisebriedenhann replied:
👍 bb6d6fa7feff8b19424f7548f74363d1f349fdd2

zubin commented on README.md:
````diff
@@ -0,0 +1,65 @@
+As a player, I want to be able to play again with a different word so that I can play many times
+
+```sh
+hangman.start
````
Is this intended as shell (`sh`) code or ruby code?

marlisebriedenhann replied:
👍 bb6d6fa7feff8b19424f7548f74363d1f349fdd2

zubin commented on README.md:
```diff
@@ -0,0 +1,65 @@
+
+This is a ruby application that mimics the game hangman.  
+
+
```
What do you think about adding a section showing how to play, which might begin like this:

```suggestion
## How to play

```sh
$ cd path/to/hangman
$ script/hangman
Welcome to Hangman!
Press Enter to begin
# insert pretend game interaction here
```
```

marlisebriedenhann replied:
👍 bb6d6fa7feff8b19424f7548f74363d1f349fdd2

jahseng-lee replied:
As a small note, it's common practice to have `./bin/[start-script].sh` as the directory.
Not a big deal, but you'll find in popular non-web Ruby applications that `bin` is usually the directory for that script 🙂

marlisebriedenhann committed:
Refactor README
- change main heading to be application name
- change code blocks from sh to ruby
- Add How to play section
- change lives_left to lives_remaining

jahseng-lee commented:
@marlisebriedenhann really good start, can't wait to see the first PR for the User Story (Test Driven of course 😉)
